adapter/stackdriver/stackdriver.go

```diff
@@ -0,0 +1,312 @@
+	case *config.Params_AppCredentials:
+		// When using default app credentials the SDK handles everything for us.
+	}
+	if cfg.Endpoint != "" {
+		opts = append(opts, gapiopts.WithEndpoint(cfg.Endpoint))
+	}
+	return
+}
+
+// NewMetricsAspect provides an implementation for adapter.MetricsBuilder.
+func (f *factory) NewMetricsAspect(env adapter.Env, c adapter.Config, metrics map[string]*adapter.MetricDefinition) (adapter.MetricsAspect, error) {
+	cfg := c.(*config.Params)
+	types := make(map[string]sdinfo, len(metrics))
+	for name, def := range metrics {
+		info, found := cfg.MetricInfo[name]
+		if !found {
+			env.Logger().Warningf("No stackdriver info found for metric %s, skipping it", name)
+			continue
+		}
+		// TODO: do we want to make sure that the definition conforms to stackdrvier requirements? Really that needs to happen during config validation
+		types[name] = sdinfo{
+			ttype: metricType(def.Name),
+			kind:  info.Kind,
+			value: info.Value,
+		}
+	}
+
+	// Per the documentation on config.proto, if push_interval is zero we'll default to a 5s push interval
+	if cfg.PushInterval == time.Duration(0) {
+		cfg.PushInterval = 1 * time.Second
+	}
+	f.once.Do(func() {
+		f.t = time.NewTicker(cfg.PushInterval)
+	})
+
+	var err error
+	var client *monitoring.MetricClient
+	// TODO: in theory this client could live in the factory and be shared amongst many adapter instances
+	if client, err = f.createClient(cfg); err != nil {
+		return nil, env.Logger().Errorf("Failed to construct stackdriver client: %v", err)
+	}
+
+	s := &sd{
+		env:         env,
+		projectID:   cfg.ProjectId,
+		client:      client,
+		pushMetrics: client.CreateTimeSeries,
+		metricInfo:  types,
+		m:           sync.Mutex{},
+	}
+	s.startTicker(env, f.t)
+	return s, nil
+}
+
+// Extracted from NewMetricsAspect for testing - it shouldn't be possible to get an Env any other time.
+func (s *sd) startTicker(env adapter.Env, ticker *time.Ticker) {
+	env.ScheduleDaemon(func() {
+		for range ticker.C {
+			s.pushData()
+		}
+	})
+}
+
+func (s *sd) Record(vals []adapter.Value) error {
+	s.env.Logger().Infof("stackdriver.Record called with %d vals", len(vals))
+
+	// TODO: len(vals) is constant for config lifetime, consider pooling
+	data := make([]*monitoringpb.TimeSeries, 0, len(vals))
+	for _, val := range vals {
+		info, found := s.metricInfo[val.Definition.Name]
+		if !found {
+			// We weren't configured with stackdriver data about this metric, so we don't know how to publish it.
+			s.env.Logger().Infof("Skipping metric %s due to not being configured with stackdriver info about it.", val.Definition.Name)
+			continue
+		}
+
+		start, _ := ptypes.TimestampProto(val.StartTime)
+		end, _ := ptypes.TimestampProto(val.EndTime)
+		data = append(data, &monitoringpb.TimeSeries{
+			Metric: &metricpb.Metric{
+				Type:   info.ttype,
+				Labels: toStringMap(val.Labels),
+			},
+			// TODO: handle MRs; today we publish all metrics to SD's global MR because it's easy.
```

Comment:
this, to me, seems like potentially the biggest TODO in this impl. Because this also might imply config changes in a few places.